runtime/src/main/cpp/CyclicCollector.cpp

```diff
@@ -0,0 +1,472 @@
+    return nullptr;
+  }
+
+  void gcProcessor() {
+     {
+       Locker locker(&lock_);
+       KStdDeque<ObjHeader*> toVisit;
+       KStdUnorderedSet<ObjHeader*> visited;
+       int restartCount = 0;
+       while (!terminateCollector_) {
+         CHECK_CALL(pthread_cond_wait(&cond_, &lock_), "Cannot wait collector condition")
+         if (!shallRunCollector_) continue;
+         atomicSet(&gcRunning_, 1);
+         restartCount = 0;
+        restart:
+         if (restartCount > 10) {
+           COLLECTOR_LOG("wait for some time to avoid GC trashing\n");
+           struct timeval tv;
+           struct timespec ts;
+           long long nsDelta = 1000LL * 1000LL;
+           ts.tv_nsec = (tv.tv_usec * 1000LL + nsDelta) % 1000000000LL;
+           ts.tv_sec = tv.tv_sec;
+           pthread_cond_timedwait(&cond_, &lock_, &ts);
+         }
+         atomicSet(&mutatedAtomics_, 0);
+         visited.clear();
+         toVisit.clear();
+         sideRefCounts_.clear();
+         for (auto* root: rootset_) {
+           // We only care about frozen values here, as only they could become part of shared cycles.
+           if (!root->container()->frozen()) continue;
+           COLLECTOR_LOG("process root %p\n", root);
+           traverseObjectFields(root, [&toVisit, &visited](ObjHeader** location) {
+             ObjHeader* ref = *location;
+             if (ref != nullptr && visited.count(ref) == 0) {
+               toVisit.push_back(ref);
+               COLLECTOR_LOG("adding %p for visiting\n", ref);
+             }
+           });
+           if (atomicGet(&mutatedAtomics_) != 0) {
+             COLLECTOR_LOG("restarted during visiting collect\n")
+             restartCount++;
+             goto restart;
+           }
+           while (toVisit.size() > 0)  {
+             auto* obj = toVisit.front();
+             toVisit.pop_front();
+             int increment = isEdgeCounted(obj) ? 1 : 0;
+             obj = clearEdgeIsCounted(obj);
+             COLLECTOR_LOG("visit %p\n", obj);
+             auto* objContainer = obj->container();
+             if (objContainer == nullptr) continue;  // Permanent object.
+             if (!objContainer->frozen()) continue;
+             sideRefCounts_[obj] += increment;
+             if (atomicGet(&mutatedAtomics_) != 0) {
+               COLLECTOR_LOG("restarted during rootset visit\n")
+               restartCount++;
+               goto restart;
+             }
+             if (visited.count(obj) == 0) {
+               visited.insert(obj);
+               traverseObjectFields(obj, [&toVisit, &visited, obj](ObjHeader** location) {
+                 ObjHeader* ref = *location;
+                 if (ref != nullptr) {
+                   // We shall not account for edges inside the same frozen container, unless it originates
+                   // from an atomic reference.
+                   bool shallCount = isAtomicReference(obj) || (obj->container() != ref->container());
```

Comment:
Why not increment `sideRefCounts_` here instead?
(Of course after placing roots to `toVisit` instead of traversing their fields [here](https://github.com/JetBrains/kotlin-native/pull/3742/files#diff-2f35eb64c8651572916c7aeb9cea20d7R199)).
This seems to be simpler.